### DISCUSSION TITLE
Is type variable supported for the dispatch?

### DISCUSSION THREAD

**Discussion by snwnde** [coady/multimethod#104] [Q&A] [answered]
Consider the defined type variables,
```python
K = TypeVar('K')
V = TypeVar('V')
```

and a custom generic dictionary,
```python
class MyGenericDict(collections.UserDict[K, V]):
    ...
```

I am trying to define,
```python
class MyGenericDict(collections.UserDict[K, V]):
    
    @multimethod.multimethod
    def __mul__(self, other: MyGenericDict[K, Any]):
        return self.__class__({key: value * other[key] for key, value in self.items()})

    @multimethod.multimethod
    def __mul__(self, other: V):
        return self.__class__({key: value * other for key, value in self.items()})
```

It seems that as long as `other` is not an instance of `MyGenericDict`, the call is dispatched to the second definition, no matter if it is an instance of the actual concrete `V`.

Is this a limitation of the typing system in python, or that of the `multimethod` package? Or have I done something wrong?

**Comment by coady** [comment, marked as answer]
There are no constraints on `TypeVar('V')`, so it matches anything.

**Reply by snwnde** [reply]
Thanks. Is there way to match `int` for `MyGenericDict[str, int]`, `float` for `MyGenericDict[str, float]` and any type `V` for `MyGenericDict[str, V]` or not? From what I understood it seems impossible since `V` cannot be bound.

**Reply by coady** [reply]
`V` can be bound: `TypeVar('V', int, float)`.

Or `MyGenericDict[str, int]` and `MyGenericDict[str, float]` can be used.

**Reply by snwnde** [reply]
Well, this would not work for generally unbound `V`, which could be `int`, `float`, another parameterisation of `MyGenericDict` or anything else. That said, this design might just be bad and I feel this should not be the way to go. I need to think more. Thank you for your answers!
